zone/mob.cpp

```diff
@@ -593,57 +593,48 @@ void Mob::CalcSeeInvisibleLevel()
 	see_invis = std::max({ spellbonuses.SeeInvis, itembonuses.SeeInvis, aabonuses.SeeInvis, innate_see_invis });
 }
 
-void Mob::CalcInvisibleLevel()
-{
-	bool is_invisible = invisible;
+bool Mob::HasAnInvisibilityEffect() {
+	return invisible || hidden || improved_hidden || invisible_animals || invisible_undead;
+}
+
+void Mob::BreakCharmPetIfConditionsMet() {
+	auto pet = GetPet();
+	if (pet && pet->GetPetType() == petCharmed && HasAnInvisibilityEffect()) {
+		if (RuleB(Pets, LivelikeBreakCharmOnInvis) || IsInvisible(pet)) {
+			pet->BuffFadeByEffect(SE_Charm);
+		}
+		LogRules("Pets:LivelikeBreakCharmOnInvis for [{}] | Invis [{}] - Hidden [{}] - Shroud of Stealth [{}] - IVA [{}] - IVU [{}]",
+				 GetCleanName(), invisible, hidden, improved_hidden, invisible_animals, invisible_undead);
+	}
+}
+
+void Mob::CalcInvisibleLevel() {
+	bool was_invisible = invisible;
 
 	invisible = std::max({ spellbonuses.invisibility, nobuff_invisible });
 	invisible_undead = spellbonuses.invisibility_verse_undead;
 	invisible_animals = spellbonuses.invisibility_verse_animal;
 
-	if (!is_invisible && invisible) {
-		SetInvisible(Invisibility::Invisible, true);
-		return;
-	}
-
-	if (is_invisible && !invisible) {
+	if (was_invisible != invisible) {
```

Comment:
Adjusted formatting here

Try to not use `else` if we can. Prevents code getting needlessly nested in the long term of maintenance.

Adjusted to

```cpp
void Mob::CalcInvisibleLevel()
{
	bool was_invisible = invisible;

	invisible         = std::max({spellbonuses.invisibility, nobuff_invisible});
	invisible_undead  = spellbonuses.invisibility_verse_undead;
	invisible_animals = spellbonuses.invisibility_verse_animal;

	if (was_invisible != invisible) {
		SetInvisible(invisible, true);
		return;
	}

	BreakCharmPetIfConditionsMet();
}
```